compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -246,11 +248,135 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                 self.safety_context = closure_visitor.safety_context;
             }
+            ExprKind::Borrow { borrow_kind, arg } => {
+                let arg = match self.thir[arg].kind {
+                    ExprKind::Scope { value, .. } => value,
+                    _ => arg,
+                };
+                if let ExprKind::Field { lhs, .. } = self.thir[arg].kind {
+                    if let ty::Adt(adt_def, _) = self.thir[lhs].ty.kind() {
+                        if (Bound::Unbounded, Bound::Unbounded)
+                            != self.tcx.layout_scalar_valid_range(adt_def.did)
+                        {
+                            match borrow_kind {
+                                BorrowKind::Shared | BorrowKind::Unique
+                                    if !self.thir[arg].ty.is_freeze(
+                                        self.tcx.at(self.thir[arg].span),
+                                        self.param_env,
+                                    ) =>
+                                {
+                                    self.requires_unsafe(expr.span, BorrowOfLayoutConstrainedField);
+                                }
+                                BorrowKind::Shallow | BorrowKind::Mut { .. } => {
+                                    self.requires_unsafe(
+                                        expr.span,
+                                        MutationOfLayoutConstrainedField,
+                                    );
+                                }
+                                _ => {}
+                            }
+                        }
+                    }
+                }
+            }
+            ExprKind::Assign { lhs, .. } | ExprKind::AssignOp { lhs, .. } => {
+                let lhs = match self.thir[lhs].kind {
+                    ExprKind::Scope { value, .. } => value,
+                    _ => lhs,
+                };
+                if let ExprKind::Field { lhs, .. } = self.thir[lhs].kind {
+                    if let ty::Adt(adt_def, _) = self.thir[lhs].ty.kind() {
+                        if (Bound::Unbounded, Bound::Unbounded)
+                            != self.tcx.layout_scalar_valid_range(adt_def.did)
+                        {
+                            self.requires_unsafe(expr.span, MutationOfLayoutConstrainedField);
+                        }
+                    }
+                }
+            }
             _ => {}
         }
 
         visit::walk_expr(self, expr);
     }
+
+    fn visit_pat(&mut self, pat: &Pat<'tcx>) {
+        match &*pat.kind {
+            PatKind::Variant { .. } | PatKind::Leaf { .. } => {
+                if let ty::Adt(adt_def, ..) = pat.ty.kind() {
+                    if (Bound::Unbounded, Bound::Unbounded)
+                        != self.tcx.layout_scalar_valid_range(adt_def.did)
+                    {
+                        check_for_ref_bindings(self, pat);
+
+                        fn check_for_ref_bindings<'a, 'tcx>(
```

Comment:
I don't see why does this only recurses on certain kinds of patterns, what am I missing? Also, it's a bit of a shame to re-implement visiting, and we visit some patterns twice. Can we implement this by having a field in `UnsafetyVisitor` track when we are in a layout-restricted-type pattern, and let `walk_pat` do its thing?

Reply:
I have implemented your suggestion now, although with the same reservation about `Deref` as below.